cmd/pitr/recoverer/recoverer.go

```diff
@@ -0,0 +1,258 @@
+	}
+	err = r.setBinlogs(startGTID)
+	if err != nil {
+		return errors.Wrap(err, "get binlog list")
+	}
+	err = r.recover()
+	if err != nil {
+		return errors.Wrap(err, "recover")
+	}
+
+	return nil
+}
+
+func (r *Recoverer) recover() error {
+	host, err := r.getHost()
+	if err != nil {
+		return errors.Wrap(err, "get host")
+	}
+
+	flags := ""
+	endTime := time.Time{}
+	// TODO: add logic for all types
+	switch r.recoverType {
+	case Skip:
+		flags = " --exclude-gtids=" + r.gtidSet
+	case Transaction:
+	case Date:
+		flags = ` --stop-datetime="` + r.recoverTime + `"`
+
+		format := "2006-01-02 15:04:05"
+		t, err := time.Parse(format, r.recoverTime)
+		if err != nil {
+			return errors.Wrap(err, "parse date")
+		}
+		endTime = t
+	case Latest:
+	default:
+		return errors.New("wrong recover type")
+	}
+
+	for _, binlog := range r.binlogs {
+		log.Println("working with", binlog)
+
+		if r.recoverType == Date {
+			binlogArr := strings.Split(binlog, ":")
+			if len(binlogArr) < 2 {
+				return errors.New("get timestamp from binlog name")
+			}
+			binlogTime, err := strconv.ParseInt(binlogArr[1], 10, 64)
+			if err != nil {
+				return errors.Wrap(err, "get binlog time")
+			}
+			if binlogTime > endTime.Unix() {
+				return nil
+			}
+		}
+
+		binlogObj, err := r.storage.GetObject(binlog)
+		if err != nil {
+			return errors.Wrap(err, "get obj")
+		}
+
+		cmdString := "cat | mysqlbinlog" + flags + " - | mysql -h" + host + " -u" + r.pxcUser + " -p$PXC_PASS"
+		cmd := exec.Command("sh", "-c", cmdString)
+
+		cmd.Stdin = binlogObj
+		var outb, errb bytes.Buffer
+		cmd.Stdout = &outb
+		cmd.Stderr = &errb
+		err = cmd.Run()
+		if err != nil {
+			return errors.Wrap(err, "cmd run")
+		}
+
+		if errb.Bytes() != nil {
+			log.Println(errors.Errorf("cmd error: %s, stdout: %s", errb.String(), outb.String()))
+		}
+	}
+
+	return nil
+}
+
+func (r *Recoverer) getLastBackupGTID() (int64, error) {
+	sep := []byte("GTID of the last")
+
+	infoObj, err := r.storage.GetObject(r.backupName + "/xtrabackup_info.lz4.00000000000000000000")
+	if err != nil {
+		return 0, errors.Wrap(err, "get object")
+	}
+	content, err := ioutil.ReadAll(infoObj)
+	if err != nil {
+		return 0, errors.Wrap(err, "read object")
+	}
+	startIndex := bytes.Index(content, sep)
+	if startIndex == -1 {
+		return 0, errors.New("no gtid data in backup")
+	}
+	newOut := content[startIndex+len(sep):]
+	e := bytes.Index(newOut, []byte("'\n"))
+	if e == -1 {
+		return 0, errors.New("cant find gtid data in backup")
+	}
+
+	set := newOut[:e]
+	setArr := bytes.Split(set, []byte("-"))
+	if len(setArr) < 2 {
+		return 0, errors.New("cant find lastgtid in backup")
+	}
+	idBytes := setArr[len(setArr)-1]
+
+	id, err := strconv.ParseInt(string(idBytes), 10, 64)
+	if err != nil {
+		return 0, errors.New("cant convert last gtid to int64")
+	}
+
+	return id, nil
+}
+
+func (r *Recoverer) setBinlogs(startID int64) error {
+	saveBinlog := false
+	for _, binlog := range r.storage.ListObjects("") {
```

Comment:
we sent `""` as path prefix here, It is critical to work inside separate directories in one bucket.
e.g. cluster1 should store binlogs into `s3://operator-testing/cluster1` and cluster2 into `s3://operator-testing/cluster2`

Reply:
https://github.com/percona/percona-xtradb-cluster-operator/blob/K8SPXC-530/cmd/pitr/recoverer/recoverer.go#L241

Reply:
https://github.com/percona/percona-xtradb-cluster-operator/pull/673/commits/e32dc99092f74f4a3a93b9648e3cbde200ccf9c0